README
DeepZoom Images
===============

DZI - now a Mirkosoft acquired technology - is a way to view
a very large image in different resolutions over the web. This
is done with the help of an image pyramid.


INSTALLATION

To install this module, run the following commands:

	perl Build.PL
	./Build
	./Build test
	./Build install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Graphics::DZI

You can also look for information at:

    RT, CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Graphics-DZI

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Graphics-DZI

    CPAN Ratings
        http://cpanratings.perl.org/d/Graphics-DZI

    Search CPAN
        http://search.cpan.org/dist/Graphics-DZI


COPYRIGHT AND LICENCE

Copyright (C) 2010 Robert Barta

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.



Changes
Revision history for Graphics-DZI

!! means break of backward compatibility

0.04    Thu Mar 18 20:35:25 CET 2010
	- fixed the problem of script not being installed automatically

0.03    Thu Mar 18 19:20:25 CET 2010
	- added sparse image support via overlays
	- added sparse document images
	- added Graphics::DZI: overlays option
	- changed invocation line in deepzoom
	- changed defaults for tilesizes to 256

0.02	Sun Mar 14 18:13:34 CET 2010
	- first release

0.01    Tue Mar  9 17:44:54 CET 2010
	- experimental version



Build.PL
use strict;
use warnings;
use Module::Build;


use Module::Build;
my $class = Module::Build->subclass(
    class => "Module::Build::Custom",
    code => <<'SUBCLASS' );

sub ACTION_upload {
    my $self = shift;
    my $v = $self->dist_version();
    $self->do_system(('cpan-upload', "Graphics-DZI-$v.tar.gz"));
}
SUBCLASS

my $builder = Module::Build::Custom->new(
    module_name         => 'Graphics::DZI',
    license             => 'perl',
#    version             => $version,
    dist_author         => 'Robert Barta <drrho@cpan.org>',
    dist_version_from   => 'lib/Graphics/DZI.pm',
    build_requires => {
	'Moose'               => undef,
	'Image::Magick'       => undef,
	'File::Path'          => undef,
	'File::Slurp'         => undef,
	'File::Basename'      => undef,
	'Getopt::Long'        => undef,
	'Pod::Usage'          => undef,
	'Log::Log4perl'       => undef,
	'Image::Magick'       => undef,
	'List::MoreUtils'     => undef,
        'Test::More'          => undef,
	'Test::Exception'     => undef,
	'Test::Pod::Coverage' => undef,
    },
    script_files       => [ 'script/deepzoom' ],
    add_to_cleanup     => [ 'Graphics-DZI-*' ],
    create_makefile_pl => 'traditional',
);

$builder->create_build_script();

__END__



Makefile.PL
# Note: this file was auto-generated by Module::Build::Compat version 0.340201
use ExtUtils::MakeMaker;
WriteMakefile
(
          'NAME' => 'Graphics::DZI',
          'VERSION_FROM' => 'lib/Graphics/DZI.pm',
          'PREREQ_PM' => {
                           'File::Basename' => undef,
                           'File::Path' => undef,
                           'File::Slurp' => undef,
                           'Getopt::Long' => undef,
                           'Image::Magick' => undef,
                           'List::MoreUtils' => undef,
                           'Log::Log4perl' => undef,
                           'Moose' => undef,
                           'Pod::Usage' => undef,
                           'Test::Exception' => undef,
                           'Test::More' => undef,
                           'Test::Pod::Coverage' => undef
                         },
          'INSTALLDIRS' => 'site',
          'EXE_FILES' => [
                           'script/deepzoom'
                         ],
          'PL_FILES' => {}
        )
;


lib/Graphics/DZI/A4.pm
package Graphics::DZI::A4;

use warnings;
use strict;

use Moose;
extends 'Graphics::DZI::Files';

our $log;
use Log::Log4perl;
BEGIN {
    $log = Log::Log4perl->get_logger ();
}

=head1 NAME

Graphics::DZI::A4 - DeepZoom Image Pyramid Generation, specifically for documents

=head1 SYNOPSIS

    use Graphics::DZI::A4;
    $Graphics::DZI::log    ->level ($loglevel);
    $Graphics::DZI::A4::log->level ($loglevel);
    my $dzi = new Graphics::DZI::A4 (A4s      => \@images,
				     overlap  => $overlap,
				     tilesize => $tilesize,
                                     path     => './',
                                     prefix   => 'xxx',
				     'format' => $format,
                             );
    use File::Slurp;
    write_file ('xxx.xml', $dzi->descriptor );
    $dzi->iterate ();

=head1 DESCRIPTION

This subclass of L<Graphics::DZI::Files> is specifically though for images covering document
pages. While it is named C<A4>, this is mostly historical; as long as all your images have the same
dimensions, this package should.

The idea is that the whole document (the set of images) forms a large image, the individual images
organized in a square fashion (1x1, 2x2, 4x4, ...). At the highest zoom level of course all pages
will be visible. But if you zoom out, then not only the pages get smaller. Also the pages shown will
be reduced, so that at the smallest zoom level only the first page is visible.

=head1 INTERFACE

=head2 Constructor

Other than the superclass L<Graphics::DZI::Files> this class takes an array (reference) to a list of
images.

=over

=item C<A4s> (no default, list reference)

Do not be fooled by the A4; any format should do.

=back

=cut

has '+image'    => (isa => 'Image::Magick', required => 0);
has 'A4s'       => (isa => 'ArrayRef',      is => 'ro'    );
has 'W'         => (isa => 'Int'   ,        is => 'rw');
has 'H'         => (isa => 'Int'   ,        is => 'rw');
has 'sqrt'      => (isa => 'Num',           is => 'rw');

sub BUILD {
    my $self = shift;
    warn "in BUILD";
    my ($W, $H) = $self->A4s->[0]->GetAttributes ('width', 'height');     # single A4
    $self->W ($W);
    $self->H ($H);

    use POSIX;
    my $log2 = POSIX::ceil (log (scalar @{$self->A4s}) / log (2));        # next fitting 2-potenz
    $log2++ if $log2 % 2;                                                 # we can only use even ones
    $self->sqrt ( 2**($log2/2) );                                         # how many along one edge when we organize them into a square?
}

=head2 Methods

=over

=item B<iterate>

This iterate honors the fact that we are dealing with a set of documents, not ONE large image.

=cut

sub _list2huge {
    my $sqrt = shift;
    my ($W, $H) = (shift, shift);

    my $dim = sprintf "%dx%d", map { $_ * $sqrt } ($W, $H);
    $log->debug ("building composite document: DIM $dim ($sqrt)");
    use Image::Magick;
    my $huge = Image::Magick->new ($dim);
    $huge->Read ('xc:white');

    foreach my $a (0 .. $sqrt*$sqrt - 1) {
	my ($i, $j) = ( int( $a / $sqrt)  , $a % $sqrt );
	$log->debug ("    index $a (x,y) = $i $j");

	$huge->Composite (image => $_[$a],
			  x     => $i * $W,
			 'y'    => $j * $H,
			  compose => 'Over',
	    );
    }
#    $huge->Display();
    return $huge;
}


sub iterate {
    my $self = shift;

    my $overlap_tilesize = $self->tilesize + 2 * $self->overlap;
    my $border_tilesize  = $self->tilesize +     $self->overlap;

    $self->image ( _list2huge ($self->sqrt, $self->W, $self->H, @{ $self->A4s }) );
    my ($WIDTH, $HEIGHT) = $self->image->GetAttributes ('width', 'height');
    $log->debug ("total dimension: $WIDTH, $HEIGHT");
    use POSIX;
    my $MAXLEVEL = POSIX::ceil (log ($WIDTH > $HEIGHT ? $WIDTH : $HEIGHT) / log (2));
    $log->debug ("   --> $MAXLEVEL");

    my ($width, $height) = ($WIDTH, $HEIGHT);
    foreach my $level (reverse (0..$MAXLEVEL)) {

	my ($x, $col) = (0, 0);
	while ($x < $width) {
	    my ($y, $row) = (0, 0);
	    my $tile_dx = $x == 0 ? $border_tilesize : $overlap_tilesize;
	    while ($y < $height) {

		my $tile_dy = $y == 0 ? $border_tilesize : $overlap_tilesize;

		my $tile = $self->crop (1, $x, $y, $tile_dx, $tile_dy);         # scale is here always 1
		$self->manifest ($tile, $level, $row, $col);

		$y += ($tile_dy - 2 * $self->overlap);
		$row++;
	    }
	    $x += ($tile_dx - 2 * $self->overlap);
	    $col++;
	}
	($width, $height) = map { int ($_ / 2) } ($width, $height);             # half size, and remember this is A4!

	$self->sqrt ($self->sqrt / 2)                                           # FUCKINGLY RETARDED OO-crap
	    if $self->sqrt > 1;                                                 # we do not go below 0
	$self->image ( _list2huge ($self->sqrt,                                 # pack sqrt x sqrt A4s into one image
				   $self->W, $self->H,
				   @{ $self->A4s }
		       ) );
	$self->image->Resize (width => $width, height => $height);            # at higher levels we need to resize that properly
    }
}

=item B<descriptor>

Also the descriptor generation is a bit special.

=cut

sub descriptor {
    my $self = shift;
    my $overlap  = $self->overlap;
    my $tilesize = $self->tilesize;
    my $format   = $self->format;
    my ($width, $height) = map { $_ * $self->sqrt }  ($self->W, $self->H);
    return qq{<?xml version='1.0' encoding='UTF-8'?>
<Image TileSize='$tilesize'
       Overlap='$overlap'
       Format='$format'
       xmlns='http://schemas.microsoft.com/deepzoom/2008'>
    <Size Width='$width' Height='$height'/>
</Image>
};


}

=back

=head1 AUTHOR

Robert Barta, C<< <drrho at cpan.org> >>

=head1 COPYRIGHT & LICENSE

Copyright 2010 Robert Barta, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut

our $VERSION = '0.02';

"against all odds";


lib/Graphics/DZI/Files.pm
package Graphics::DZI::Files;

use warnings;
use strict;

use Moose;
extends 'Graphics::DZI';

our $log;
use Log::Log4perl;
BEGIN {
    $log = Log::Log4perl->get_logger ();
}

=head1 NAME

Graphics::DZI::Files - DeepZoom Image Pyramid Generation, File-based

=head1 SYNOPSIS

  use Graphics::DZI::Files;
  my $dzi = new Graphics::DZI::Files (image    => $image,
				      overlap  => 4,
				      tilesize => 256,
				      scale    => 2,
 				      format   => 'png',
                                      prefix   => 'xxx',
                                      path     => '/where/ever/');
  use File::Slurp;
  write_file ('/where/ever/xxx.xml', $dzi->descriptor);
  $dzi->iterate ();

=head1 DESCRIPTION

This subclass of L<Graphics::DZI> generates tiles and stores them at the specified path location.

=head1 INTERFACE

=head2 Constructor

Additional to the parent class L<Graphics::DZI>, the constructor takes the following fields:

=over

=item C<format> (default C<png>):

An image format (C<png>, C<jpg>, ...). Any format L<Image::Magick> understands will do.

=item C<path>:

A directory name (including trailing C</>) where the tiles are written to.

=item C<prefix>:

The string to be prefixed the C<_files/> part in the directory name. Usually the name of the image
to be converted. No slashes.

=back

=cut

#has 'format'   => (isa => 'Str'   ,        is => 'ro', default => 'png');
has 'path'    => (isa => 'Str'   ,        is => 'ro');
has 'prefix'  => (isa => 'Str'   ,        is => 'ro');

=head2 Methods

=over

=item B<manifest>

This method writes any tile to a file, appropriately named for DZI inclusion.

=cut

sub manifest {
    my $self  = shift;
    my $tile  = shift;
    my $level = shift;
    my $row   = shift;
    my $col   = shift;

    my $path = $self->path . "$level/";
    use File::Path qw(make_path);
    make_path ($path);

    my $filen = $path . (sprintf "%s_%s", $col, $row ) . '.' . $self->format;
    $log->debug ("saving tile $level $row $col --> $filen");
    $tile->Write( $filen );
}

=back

=head1 AUTHOR

Robert Barta, C<< <drrho at cpan.org> >>

=head1 COPYRIGHT & LICENSE

Copyright 2010 Robert Barta, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut

our $VERSION = '0.02';
"against all odds";


lib/Graphics/DZI/Overlay.pm
package Graphics::DZI::Overlay;

use strict;
use warnings;

use Moose;

our $log;
use Log::Log4perl;
BEGIN {
    $log = Log::Log4perl->get_logger ();
}

=head1 NAME

Graphics::DZI::Overlay - DeepZoom Image Pyramid, Sparse Images

=head1 SYNOPSIS

  # build some overlays first
  use Graphics::DZI::Overlay;
  my $o1 = new Graphics::DZI::Overlay (image => ...,       # what is the image?
                                       x => 1000, y, 1000, # where on the canvas?
                                       squeeze => 64);     # how much smaller than the canvas?
  my $o2 = new Graphics::DZI::Overlay (image => ...,       # what is the image?
                                       x => 2000, y, 2000, # where on the canvas?
                                       squeeze => 32);     # how much smaller than the canvas?
  # then add the overlay over the canvas
  use Graphics::DZI::Files;
  my $dzi = new Graphics::DZI::Files (image    => $image,
                                      overlap  => 4,
                                      tilesize => 512,
                                      format   => 'png',
                                      overlays => [ $o1, $o2],
                                      path     => $path . 'xxx_files/',
                                      prefix   => 'xxx',
                                      );
   # normal DZI generation
   write_file ($path . 'xxx.xml', $dzi->descriptor);
   $dzi->iterate ();

=head1 DESCRIPTION

This package can hold one overlay image, together with a coordinate and a factor how much this
images is smaller than the canvas onto which the image is to be put.

=head1 INTERFACE

=head2 Constructor

It expects the following fields:

=over

=item C<image>: (required)

L<Image::Magick> object.

=item C<x>,C<y>: (integers, no default)

Coordinates of the top-left corner of the above image on the canvas 

=item C<squeeze>: (integers, no default)

A factor how much the image should be made smaller, relative to the canvas. I use a power of two to
avoid that the canvas is a bit fuzzy.

=back

=cut

has 'image'   => (isa => 'Image::Magick', is => 'rw', required => 1);
has 'x'       => (isa => 'Int', is => 'rw');
has 'y'       => (isa => 'Int', is => 'rw');
has 'squeeze' => (isa => 'Num', is => 'rw');

=head2 Methods

=over

=item B<halfsize>

Makes the overlay smaller by 2. This will be called by the DZI algorithm.

=cut

sub halfsize {
    my $self = shift;
    my ($w, $h) = $self->image->GetAttributes ('width', 'height');                     # current dimensions
    $self->image->Resize (width => int($w/2), height => int($h/2));                    # half size
    $self->{x} /= 2;                                                                   # dont forget x, y 
    $self->{y} /= 2;
}

=item B<crop>

Gets a tile off the overlay.

=cut

sub crop {
    my $self = shift;
    my ($tx, $ty, $tdx, $tdy) = @_;

    my ($w, $h) = $self->{image}->GetAttributes ('width', 'height');
    $self->{dx} = $w;
    $self->{dy} = $h;

#    warn "before intersection tile $tile"; $tile->Display() if $tile;
    if (my $r = _intersection ($tx,        $ty,        $tx+$tdx,                 $ty+$tdy,                   # tile and overlay intersect?
			       $self->{x}, $self->{y}, $self->{x} + $self->{dx}, $self->{y} +$self->{dy})) {
#	    warn " intersection!";
	my ($ox, $oy, $dx, $dy) = (
	    $r->[0] - $self->{x},                                                      # x relative to overlay
	    $r->[1] - $self->{y},                                                      # y relative to overlay

	    $r->[2] - $r->[0],                                                         # width of the intersection
	    $r->[3] - $r->[1],                                                         # height
	    );

	my $oc = $self->{image}->clone;
#	warn "overlay clone "; $oc->Display();
	$oc->Crop (geometry => "${dx}x${dy}+${ox}+${oy}");
#	warn "cropped oc";   $oc->Display();

#	unless ($tile) {                                                               # this just makes sure that we are composing onto SOMETHING
##	    warn "XXXXXXXXX generating substitute tile";
	my $tile = Image::Magick->new ("${tdx}x${tdy}");                               # create an empty one
	$tile->Read ('xc:yellow');                                                 # paint it white (otherwise composite would not work?)
	$tile->Transparent (color => 'yellow');
#	warn "substitute tile "; $tile->Display();
#	}
#	warn "before overlay tile "; $tile->Display();
	$tile->Composite (image => $oc,
			  x     => $r->[0] - $tx,                                      # intersection left/top relative to tile
			  'y'   => $r->[1] - $ty,
			  compose => 'Over',
		);
#	warn "after overlay tile "; $tile->Display();
	return $tile;
    }
    return undef;
}

sub _intersection {
    my ($ax, $ay, $axx, $ayy,
	$bx, $by, $bxx, $byy) = @_;

    if (_intersects ($ax, $ay, $axx, $ayy,
		     $bx, $by, $bxx, $byy)) {
	return [
	    $ax  > $bx  ? $ax  : $bx,
	    $ay  > $by  ? $ay  : $by,
	    $axx > $bxx ? $bxx : $axx,
	    $ayy > $byy ? $byy : $ayy
	    ];
    }
}

sub _intersects {
    my ($ax, $ay, $axx, $ayy,
	$bx, $by, $bxx, $byy) = @_;

    return undef
	if $axx < $bx
	|| $bxx < $ax
	|| $ayy < $by
	|| $byy < $ay;
    return 1;
}

=back

=head1 AUTHOR

Robert Barta, C<< <drrho at cpan.org> >>

=head1 COPYRIGHT & LICENSE

Copyright 2010 Robert Barta, all rights reserved.

This program is free software; you can redistribute it and/or modify it under the same terms as Perl
itself.

=cut

our $VERSION = '0.01';

"against all odds";



lib/Graphics/DZI.pm
package Graphics::DZI;

use warnings;
use strict;
use POSIX;

use Moose;

our $log;
use Log::Log4perl;
BEGIN {
    $log = Log::Log4perl->get_logger ();
}

=head1 NAME

Graphics::DZI - DeepZoom Image Pyramid Generation

=head1 SYNOPSIS

  use Graphics::DZI;
  my $dzi = Graphics::DZI::A4 (image    => $image,
                               overlap  => $overlap,
                               tilesize => $tilesize,
                               format   => $format,
                               );

  write_file ('/var/www/xxx.xml', $dzi->descriptor);
  $dzi->iterate ();
  # !!! this does only display the tiles on the screen
  # !!! see Graphics::DZI::Files for a subclass which 
  # !!! actually writes to files

=head1 DESCRIPTION

This base package generates tiles from a given image in such a way that they follow the DeepZoom
image pyramid scheme. Consequently this image becomes zoomable with tools like Seadragon.

  http://en.wikipedia.org/wiki/Deep_Zoom

As this is a base class, you may want to look either at the I<deepzoom> script which operators on
the command line, or at one of the subclasses.

=head1 INTERFACE

=head2 Constructor

The constructor accepts the following fields:

=over

=item C<image>

The L<Image::Magick> object which is used as canvas.

=item C<scale> (integer, default: 1)

Specifies how much the image is stretched in the process.

=item C<overlap> (integer, default: 4)

Specifies how much the individual tiles overlap.

=item C<tilesize> (integer, default: 128)

Specifies the quadratic size of each tile.

=item C<overlays> (list reference, default: [])

An array of L<Graphics::DZI::Overlay> objects which describe how further images are supposed to be
composed onto the canvas image.

=back

=cut

has 'image'    => (isa => 'Image::Magick', is => 'rw', required => 1);
has 'scale'    => (isa => 'Int',           is => 'ro', default => 1);
has 'overlap'  => (isa => 'Int',           is => 'ro', default => 4);
has 'tilesize' => (isa => 'Int',           is => 'ro', default => 256);
has 'format'   => (isa => 'Str'   ,        is => 'ro', default => 'png');
has 'overlays' => (isa => 'ArrayRef',      is => 'rw', default => sub { [] });

=head2 Methods

=over

=item B<crop>

I<$tile> = I<$dzi>->crop (I<$scale>, I<$x>, I<$y>, I<$dx>, I<$dy>)

Given the dimensions of a tile and a current (not the original)
stretch factor this method will return a tile object.

=cut

sub crop {
    my $self = shift;
    my $scale = shift;
    my ($tx, $ty, $tdx, $tdy) = @_;

    my $image = $self->{image};
    my $tile  = $image->clone;
    if ($scale != 1) {                                                                       # if our image is not quite the total space
#	warn "new canvas tile scaled $scale";
	my ($htx, $hty, $htdx, $htdy) = map { int ($_ / $scale) }
	                                ($tx, $ty, $tdx, $tdy);                              # rescale this tile to the image dims we have
	$log->debug ("rescale $tx, $ty  -->  $htx, $hty");
	$tile->Crop   (geometry => "${htdx}x${htdy}+${htx}+${hty}");                         # cut that smaller one out
	$tile->Resize ("${tdx}x${tdy}");                                                     # and make it bigger
    } else {                                                                                 # otherwise we are happy with what we have, dimension-wise
#	warn "new canvas tile unscaled";
	$tile->Crop (geometry => "${tdx}x${tdy}+${tx}+${ty}");                               # cut one out
    }
    $log->debug ("tiled ${tdx}x${tdy}+${tx}+${ty}");
#    $tile->Display();
    return $tile;
}

=item B<dimensions>

(I<$W>, I<$H>) = I<$dzi>->dimensions ('total')

(I<$W>, I<$H>) = I<$dzi>->dimensions ('canvas')

This method computes how large (in pixels) the overall image will be. If C<canvas> is passed in,
then any overlays are ignored. Otherwise their size (with their squeeze factors) are used to blow up
the canvas, so that the overlays fit onto the canvas.

=cut

sub dimensions {
    my $self = shift;
    my $what = shift || 'total';

    if ($what eq 'total') {
	use List::Util qw(max);
	my $max_squeeze = max map { $_->squeeze } @{ $self->overlays };
	$self->{scale} = defined $max_squeeze ? $max_squeeze : 1;
	return map { $_ * $self->{scale} } $self->image->GetAttributes ('width', 'height');
    } else {
	return $self->image->GetAttributes ('width', 'height');
    }
}

=item B<iterate>

I<$dzi>->iterate

This method will generate all necessary tiles, invoking the I<manifest> method. You may want to
override that one, if you do not want the tiles to be simply displayed on screen :-) Any options
you add as parameters will be passed on to I<manifest>.

=cut

sub iterate {
    my $self = shift;

    my $overlap_tilesize = $self->{tilesize} + 2 * $self->{overlap};
    my $border_tilesize  = $self->{tilesize} +     $self->{overlap};

    my ($CWIDTH, $CHEIGHT) = $self->dimensions ('canvas');                        $log->debug ("canvas dimensions: $CWIDTH, $CHEIGHT");
    my $CANVAS_LEVEL       = POSIX::ceil (log ($CWIDTH > $CHEIGHT 
					       ? $CWIDTH 
					       : $CHEIGHT) / log (2));            $log->debug ("   --> levels: $CANVAS_LEVEL");

    my ($WIDTH, $HEIGHT)   = $self->dimensions ('total');                         $log->debug ("total dimensions: $WIDTH, $HEIGHT");
    my $MAXLEVEL           = POSIX::ceil (log ($WIDTH > $HEIGHT 
					       ? $WIDTH
					       : $HEIGHT) / log (2));             $log->debug ("   --> levels: $MAXLEVEL");

    my ($width, $height) = ($WIDTH, $HEIGHT);
    my $scale = $self->{scale};
    foreach my $level (reverse (0..$MAXLEVEL)) {

	my ($x, $col) = (0, 0);
	while ($x < $width) {
	    my ($y, $row) = (0, 0);
	    my $tile_dx = $x == 0 ? $border_tilesize : $overlap_tilesize;
	    while ($y < $height) {

		my $tile_dy = $y == 0 ? $border_tilesize : $overlap_tilesize;

		my @tiles = grep { defined $_ }                                                # only where there was some intersection
                            map {
				$_->crop ($x, $y, $tile_dx, $tile_dy);                         # and for each overlay crop it onto a tile
			    } @{ $self->overlays };                                            # look at all overlays

		if (@tiles) {                                                                  # if there is at least one overlay tile
		    my $tile = $self->crop ($scale, $x, $y, $tile_dx, $tile_dy);               # do a crop in the canvas and try to get a tile
		    map {
			$tile->Composite (image => $_, x => 0, 'y' => 0, compose => 'Over')
		    } @tiles;
		    $self->manifest ($tile, $level, $row, $col);                               # we flush it

		} elsif ($level <= $CANVAS_LEVEL) {                                            # only if we are in the same granularity of the canvas
		    my $tile = $self->crop ($scale, $x, $y, $tile_dx, $tile_dy);               # do a crop there and try to get a tile
		    $self->manifest ($tile, $level, $row, $col);                               # we flush it
		}

		$y += ($tile_dy - 2 * $self->{overlap});                                       # progress y forward
		$row++;                                                                        # also the row count
	    }
	    $x += ($tile_dx - 2 * $self->{overlap});                                           # progress x forward
	    $col++;                                                                            # the col count
	}

#-- resizing canvas (virtually only)
	($width, $height) = map { int ($_ / 2) } ($width, $height);
	if (@{ $self->overlays }) {                                                            # do we have overlays from which the scale came?
	    $scale /= 2;                                                                       # the overall magnification is to be reduced
	    foreach my $o (@{ $self->overlays }) {                                             # also resize all overlays
		$o->halfsize;
	    }
	} else {
	    # keep scale == 1
	    $self->{image}->Resize (width => $width, height => $height);                       # resize the canvas for next iteration
	}
    }
}

=item B<manifest>

I<$dzi>->manifest (I<$tile>, I<$level>, I<$row>, I<$col>)

This method will get one tile as parameter and will simply display the tile on the screen.
Subclasses which want to persist the tiles, can use the additional parameters (level, row and
column) to create file names.

=cut

sub manifest {
    my $self = shift;
    my $tile = shift;
    $tile->Display();
}

=item B<descriptor>

I<$string> = I<$dzi>->descriptor

This method returns the DZI XML descriptor as string.

=cut

sub descriptor {
    my $self     = shift;
    my $overlap  = $self->{overlap};
    my $tilesize = $self->{tilesize};
    my $format   = $self->{format};
    my ($width, $height) = $self->dimensions ('total');
    return qq{<?xml version='1.0' encoding='UTF-8'?>
<Image TileSize='$tilesize'
       Overlap='$overlap'
       Format='$format'
       xmlns='http://schemas.microsoft.com/deepzoom/2008'>
    <Size Width='$width' Height='$height'/>
</Image>
};


}

=back

=head1 TODOs

See the TODOs file in the distribution.

=head1 AUTHOR

Robert Barta, C<< <drrho at cpan.org> >>

=head1 COPYRIGHT & LICENSE

Copyright 2010 Robert Barta, all rights reserved.

This program is free software; you can redistribute it and/or modify it under the same terms as Perl
itself.

=cut

our $VERSION = '0.04';

"against all odds";
